### PR TITLE
Add check cache to MixerClientImpl.

### PR DESCRIPTION
**Pull request by chowchow316**

### CODE REVIEW

File: src/client_impl_test.cc [hunk 1]
```diff
@@ -81,6 +84,7 @@ class MixerClientImplTest : public ::testing::Test, public TransportInterface {
   std::unique_ptr<WriterImpl<CheckRequest, CheckResponse>> check_writer_;
   std::unique_ptr<WriterImpl<ReportRequest, ReportResponse>> report_writer_;
   std::unique_ptr<WriterImpl<QuotaRequest, QuotaResponse>> quota_writer_;
+  std::shared_ptr<CheckCache> check_cache_;
```
Comment (qiwzhang) [inline src/client_impl_test.cc]: no need check_cache_ here

File: src/client_impl.cc [hunk 2]
```diff
@@ -34,17 +35,24 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
-  check_transport_->Send(attributes, response,
-                         [response, on_done](const Status &status) {
-                           delete response;
-                           on_done(status);
-                         });
+  Status status = check_cache_->Check(attributes, response);
+  if (status.error_code() == Code::NOT_FOUND) {
+    check_transport_->Send(attributes, response,
+                           [response, on_done](const Status &status) {
+                             delete response;
```
Comment (qiwzhang) [inline src/client_impl.cc]: need to call SetCheckResponse.

File: src/client_impl_test.cc [hunk 3]
```diff
@@ -14,6 +14,7 @@
  */
 
 #include "include/client.h"
+#include "src/check_cache.h"
```
Comment (qiwzhang) [inline src/client_impl_test.cc]: no need this include

File: src/client_impl.cc [hunk 4]
```diff
@@ -34,17 +35,27 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
   report_transport_.reset(new ReportTransport(transport));
   quota_transport_.reset(new QuotaTransport(transport));
+  check_cache_ =
+      std::unique_ptr<CheckCache>(new CheckCache(options.check_options));
 }
 
-MixerClientImpl::~MixerClientImpl() {}
+MixerClientImpl::~MixerClientImpl() { check_cache_->FlushAll(); }
 
 void MixerClientImpl::Check(const Attributes &attributes, DoneFunc on_done) {
```
Comment (qiwzhang) [inline src/client_impl.cc]: remove "const"

File: src/client_impl.cc [hunk 5]
```diff
@@ -34,17 +35,27 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
   auto response = new CheckResponse;
-  check_transport_->Send(attributes, response,
-                         [response, on_done](const Status &status) {
-                           delete response;
-                           on_done(status);
-                         });
+  Status status = check_cache_->Check(attributes, response);
+  if (status.error_code() == Code::NOT_FOUND) {
+    std::shared_ptr<CheckCache> check_cache_copy = check_cache_;
+    check_transport_->Send(
```
Comment (qiwzhang) [inline src/client_impl.cc]: here make a copy of attribute:
attribute_copy = new Attributes;
attribute_copy->attributes.swap(&attributes.attributes);

pass the copy to lamdb,

inside the lamdb,  delete it after done

File: src/client_impl.cc [hunk 6]
```diff
@@ -42,19 +42,19 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
 
 MixerClientImpl::~MixerClientImpl() { check_cache_->FlushAll(); }
 
-void MixerClientImpl::Check(Attributes &attributes, DoneFunc on_done) {
+void MixerClientImpl::Check(const Attributes &attributes, DoneFunc on_done) {
   auto response = new CheckResponse;
   Status status = check_cache_->Check(attributes, response);
```
Comment (qiwzhang) [inline src/client_impl.cc]: it will be faster if Check() can return signature.  Not need to calculate signature again for cache miss

Comment (chowchow316) [inline src/client_impl.cc]: We also need to return the status to decide cache hit or not.

Comment (qiwzhang) [inline src/client_impl.cc]: Yes, we check_cache::Check() could use another argument to output signature.

File: src/check_cache.cc [hunk 7]
```diff
@@ -50,11 +50,11 @@ bool CheckCache::ShouldFlush(const CacheElem& elem) {
   return age >= flush_interval_in_cycle_;
 }
 
-Status CheckCache::Check(const Attributes& attributes,
-                         CheckResponse* response) {
-  string request_signature = GenerateSignature(attributes);
+Status CheckCache::Check(const Attributes& attributes, CheckResponse* response,
+                         std::string* signature) {
+  *signature = GenerateSignature(attributes);
```
Comment (qiwzhang) [inline src/check_cache.cc]: We should allow signature is nullptr.  it means the caller doesn't want it.

File: src/check_cache_test.cc [hunk 8]
```diff
@@ -75,54 +75,62 @@ TEST_F(CheckCacheTest, TestDisableCache) {
 
   ASSERT_TRUE((bool)(cache_));
   CheckResponse response;
-  EXPECT_ERROR_CODE(Code::NOT_FOUND, cache_->Check(attributes1_, &response));
+  std::string signature;
+  EXPECT_ERROR_CODE(Code::NOT_FOUND,
+                    cache_->Check(attributes1_, &response, &signature));
```
Comment (qiwzhang) [inline src/check_cache_test.cc]: if you don't care, pass in nullptr.

File: src/client_impl.cc [hunk 9]
```diff
@@ -35,17 +37,30 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
-                           on_done(status);
-                         });
+  std::string signature;
+  Status status = check_cache_->Check(attributes, response, &signature);
+  if (status.error_code() == Code::NOT_FOUND) {
+    std::shared_ptr<CheckCache> check_cache_copy = check_cache_;
+    check_transport_->Send(
+        attributes, response,
+        [check_cache_copy, signature, response, on_done](const Status &status) {
+          check_cache_copy->CacheResponse(signature, *response);
```
Comment (qiwzhang) [inline src/client_impl.cc]: if status.ok() then cache response

Comment (qiwzhang) [inline src/client_impl.cc]: Add a function to parse response to status.  call this parsing function in line 58 and 63

**Comment by geeknoid** [comment]
This seems like a good forward step. But ultimately, as implemented the cache will be completely ineffective as there are going to be some highly volatile attributes that perturb the signature.

The client library will be told explicitly the set of attributes that should be used to generate a signature. The exact set will be based on the config in use.

File: src/client_impl.cc [hunk 10]
```diff
@@ -25,6 +25,12 @@ using ::google::protobuf::util::error::Code;
 
 namespace istio {
 namespace mixer_client {
+namespace {
+Status ParseStatusFromResponse(const CheckResponse &response) {
```
Comment (qiwzhang) [inline src/client_impl.cc]: call it ParseCheckResponse

**Comment by chowchow316** [comment]
Do not check in until mixer separate quota from check.

**Comment by sebastienvas** [comment]
It seems that this breaks proxy build. I added a new presubmit

**Comment by chowchow316** [comment]
Rebased this PR, and passed jenkins presubmit.

**Comment by qiwzhang** [comment]
Replaced with another PR https://github.com/istio/mixerclient/pull/45

**Pull request closed without merging**